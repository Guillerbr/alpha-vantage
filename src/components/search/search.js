import React from 'react';

import { connect } from 'react-redux';
import { searchSymbolInfo} from '../../store/reducers/searchReducer';
import { withHandlers, withState, withProps, compose } from 'recompose';

import { Input, Button, Layout, Table } from 'antd';
const { Content } = Layout;
const { Column } = Table;

const enhance = compose(
    connect(
        ({ SearchReducer }) => ({ SearchReducer }),
        {
            searchSymbolInfoConnect: searchSymbolInfo
        }
    ),
    withProps(({ SearchReducer }) => ({
        data: SearchReducer.data,
        isFetching: SearchReducer.isFetching,
        isSuccess: SearchReducer.isSuccess,
        isError: SearchReducer.isError
    })),
    withState("inputValue", "setInputValue", ""),
    withHandlers({
        onChange: ({ setInputValue }) => event => {
            setInputValue(event.target.value)
        }
    })
);

const isEmpty = (array) => array != null && array !== undefined && array.length === 0;
const searchBoxStyle = {
  display: "flex"
};
const resultBoxStyle = {
    paddingTop: "20px"
};

export const Search = enhance(({ inputValue, data, isFetching, isSuccess, isError, onChange, searchSymbolInfoConnect }) =>
    <Layout className="layout">
    <Content style={{ padding: '0 50px', marginTop: 20, marginBottom: 20 }}>
        <div style={{ background: '#fff', padding: 24, minHeight: 280 }}>
        <div style={searchBoxStyle}>
            <span>
                Pesquise pelo nome do papel (PETR4):
            </span>
        <Input
            onChange={onChange}
            type='text'
            value={inputValue}
        />
            <span>
                &nbsp;
            </span>
        <Button shape="circle" icon="search" disabled={inputValue.length < 1} onClick={() => searchSymbolInfoConnect(inputValue)}/>
        </div>
        <div style={resultBoxStyle}>
        {
            isFetching &&
            <div>Pesquisando...</div>
        }
        {
            isError &&
            <div>Ocorreu um erro ao realizar a consulta.</div>
        }
        {
            isEmpty(data.bestMatches) &&
            <div>Nenhum item encontrado!</div>
        }
        {
            isSuccess && !isEmpty(data.bestMatches) &&
            <Table dataSource={data.bestMatches} pagination={false}>
                <Column title="Símbolo" dataIndex="1. symbol" key="1. symbol"/>
                <Column title="Empresa" dataIndex="2. name" key="2. name"/>
                <Column title="Região" dataIndex="4. region" key="4. region"/>
                <Column title="Fuso Horário" dataIndex="7. timezone" key="7. timezone"/>
                <Column title="Moeda" dataIndex="8. currency" key="8. currency"/>
            </Table>
        }
        </div>
        </div>
    </Content>
    </Layout>
);
