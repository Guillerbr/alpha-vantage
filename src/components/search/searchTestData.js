export const searchTestData = {
    "bestMatches": [
        {
            "1. symbol": "BRY",
            "2. name": "Berry Petroleum Corporation",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.5714"
        },
        {
            "1. symbol": "MPC",
            "2. name": "Marathon Petroleum Corporation",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.5714"
        },
        {
            "1. symbol": "PBR",
            "2. name": "Petr√≥leo Brasileiro S.A. - Petrobras",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.3333"
        },
        {
            "1. symbol": "OXY",
            "2. name": "Occidental Petroleum Corporation",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.3200"
        },
        {
            "1. symbol": "OAS",
            "2. name": "Oasis Petroleum Inc.",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.3077"
        },
        {
            "1. symbol": "WLL",
            "2. name": "Whiting Petroleum Corporation",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.2857"
        },
        {
            "1. symbol": "UPL",
            "2. name": "Ultra Petroleum Corp.",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.2857"
        },
        {
            "1. symbol": "APC",
            "2. name": "Anadarko Petroleum Corporation",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.2857"
        },
        {
            "1. symbol": "LPI",
            "2. name": "Laredo Petroleum Inc.",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.2424"
        },
        {
            "1. symbol": "CPE",
            "2. name": "Callon Petroleum Company",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-05",
            "8. currency": "USD",
            "9. matchScore": "0.2222"
        }
    ]
};
