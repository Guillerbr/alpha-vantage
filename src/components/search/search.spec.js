import React from 'react';

import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';

import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';

import { Search} from "./search";
import { searchTestData } from "./searchTestData";

import 'jest-dom/extend-expect';
import 'react-testing-library/cleanup-after-each';

const sagaMiddleware = createSagaMiddleware();
const middlewares = [sagaMiddleware];

const mockStore = configureStore(middlewares);

const renderWithProvider = ({ store, component } = {}) => {
    return render(
        <Provider store={store}>{component}</Provider>
    );
};

const initialState = {
    searchReducer: {
        isFetching: false,
        isFetched: false,
        isError: false,
        isSuccess: false,
        data: {}
    }
};

describe('should render search component', () => {
    it('renders search and set a value', () => {
        const store = mockStore(initialState);
        const { container, queryByTestId } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        const input = container.querySelector('input[data-modal="input-search"]');
        fireEvent.change(input, { target: { value: "PETR" } });

        expect(container).toMatchSnapshot();
        expect(queryByTestId("button-search")).toBeInTheDocument();
    });
    it('renders correctly in initial state', () => {
        const store = mockStore(initialState);
        const { queryByTestId } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        expect(queryByTestId("input-search")).toBeInTheDocument();
        expect(queryByTestId("button-search")).toBeDisabled();
    });
    it('renders fetching state', () => {
        const store = mockStore({
            searchReducer: {
                isFetching: true
            }
        });
        const { container } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        expect(container).toMatchSnapshot();
    });
    it('renders error state', () => {
        const store = mockStore({
            searchReducer: {
                isError: true
            }
        });
        const { container } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        expect(container).toMatchSnapshot();
    });
    it('renders symbol not found state', () => {
        const store = mockStore({
            searchReducer: {
                isSuccess: true,
                isFetching: false,
                data: {}
            }
        });
        const { container } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        expect(container).toMatchSnapshot();
    });
    it('renders encountered symbols state', () => {
        const store = mockStore({
            searchReducer: {
                isSuccess: true,
                isFetching: false,
                data: searchTestData
            }
        });
        const { container, queryByTestId } = renderWithProvider({
            store,
            component: (
                <Search />
            )
        });
        expect(container).toMatchSnapshot();

        expect(queryByTestId("button-symbol-load")).toBeInTheDocument();
        expect(queryByTestId("symbol-name")).toBeInTheDocument();
        expect(queryByTestId("symbol-region")).toBeInTheDocument();
        expect(queryByTestId("symbol-timezone")).toBeInTheDocument();
        expect(queryByTestId("symbol-currency")).toBeInTheDocument();
    });
});
