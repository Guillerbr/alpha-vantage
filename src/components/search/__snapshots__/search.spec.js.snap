// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render search component renders encountered symbols state 1`] = `
<div>
  <div>
    <input
      data-modal="input-search"
      data-testid="input-search"
      type="text"
      value=""
    />
    <button
      data-modal="input-search"
      data-testid="button-search"
      disabled=""
    >
      Search
    </button>
    <table>
      <tbody>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              BRY
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Berry Petroleum Corporation
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              MPC
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Marathon Petroleum Corporation
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              PBR
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Petr√≥leo Brasileiro S.A. - Petrobras
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              OXY
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Occidental Petroleum Corporation
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              OAS
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Oasis Petroleum Inc.
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              WLL
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Whiting Petroleum Corporation
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              UPL
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Ultra Petroleum Corp.
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              APC
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Anadarko Petroleum Corporation
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              LPI
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Laredo Petroleum Inc.
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
        <tr>
          <td>
            <button
              data-testid="button-symbol-load"
            >
              CPE
            </button>
          </td>
          <td
            data-testid="symbol-name"
          >
            Callon Petroleum Company
          </td>
          <td
            data-testid="symbol-region"
          >
            United States
          </td>
          <td
            data-testid="symbol-timezone"
          >
            UTC-05
          </td>
          <td
            data-testid="symbol-currency"
          >
            USD
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`should render search component renders error state 1`] = `
<div>
  <div>
    Ops! an error occured.
  </div>
</div>
`;

exports[`should render search component renders fetching state 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`should render search component renders search and set a value 1`] = `
<div>
  <div>
    <input
      data-modal="input-search"
      data-testid="input-search"
      type="text"
      value="PETR"
    />
    <button
      data-modal="input-search"
      data-testid="button-search"
    >
      Search
    </button>
    <table>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`should render search component renders symbol not found state 1`] = `
<div>
  <div>
    Symbol not found.
  </div>
</div>
`;
